source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md

```diff
@@ -0,0 +1,254 @@
+
+namespace App\Presenter;
+
+final class ProductPresenter extends AbstractBasePresenter
+{
+    /**
+     * @inject
+     * @var ProductRepository
+     */
+    public $productRepository;
+}
+```
+
+Using `@inject` annotations over constructor injection is **fast, short and it just works**.
+
+Ok, why not use it everywhere:
+
+```language-php
+// app/Repository/ProductRepository.php
+
+namespace App\Repository;
+
+class ProductRepository
+{
+    /**
+     * @inject
+     * @var Doctrine
+     */
+    public $entityManager;
+}
+```
+
+and 
+
+```language-yaml
+# app/config/config.neon
+
+services:
+    -
+        class: App\Repository\ProductRepository
+            inject: on
+```
+
+## Your Code is Seen as Manual How to Write 
+
+Why? Because "what you see is what you write". New programmer joins the teams, sees this handy @inject feature and uses when possible and handy.
+
+Some of you, who already talked about `@inject` method usage already there are some and only few specific places to use it.
+
+## Where to only @inject?
+
+**To prevent constructor hell**. If you meet this term first time, go read [this short explanation](https://phpfashion.com/di-a-predavani-zavislosti#toc-constructor-hell) from David Grudl.
+
+The best use case is `AbstractBasePresenter`.  
+Let's say I need `Translator` service in all of my presenters.
+
+```language-php
+// app\Presenter\AbstractBasePresenter.php
+
+namespace App\Presenter;
+
+abstract class AbstractBasePresenter extends Nette\Application\UI\Presenter
+{
+    /**
+     * @inject
+     * @var Translator
+     */
+    public $translator;
+}
+```
+
+And I can use it `ProductPresenter` alongside with constructor injection
+ 
+```language-php
+// app/Presenter/ProductPresenter.php
+
+namespace App\Presenter;
+
+final class ProductPresenter extends AbstractBasePresenter
+{
+    /**
+     * @var ProductRepository
+     */
+    private $productRepository;
+    
+    public function __construct(ProductRepository $productRepository) 
+    {
+        $this->productRepository = $productRepository; 
+    }
+}
+```
+
+This is quite clean and easy to use, because presenters in Nette are [injected to by default](https://github.com/nette/application/blob/3165d3a8dab876f4364cdcba450a33ab0182049a/src/Bridges/ApplicationDI/ApplicationExtension.php#L111-L116).
+ 
+## Level up
+
+But what if we have other objects that:
+ 
+ - **inherit from abstract parent**
+ - needs **1 service available everywhere**
+
+2 common case pop to my mind:
+
+- `AbstractBaseRepository` for all our repositories
+- `AbstractBaseControl` for all our components
+
+Let's take the first one:
+
+```language-php
+// app/Repository/AbstractBaseRepository.php
+
+namespace App\Repository;
+
+use Doctrine\ORM\EntityManagerInterface;
+
+abstract class AbstractBaseRepository
+{
+    /**
+     * @var EntityManagerInterface
+     */
+    protected $entityManager;
+    
+    public function setEntityManager(EntityManagerInterface $entityManager)
```

Comment:
I was hoping you'd point out why is this bad, now show more examples how to use it more easily.